### ISSUE TITLE
[homematic] wired devices not found - null pointer exception

### ISSUE THREAD

**Issue by s0170071** [openhab/openhab-addons#10945]
Since the upgrade from 3.0.1 to 3.1.0 (RPI4, piVCCU) all of my Homematic-wired things are not found any more. 
Affects Homematic-wired, Homematic-IP. RF devices work fine.
Also affects both, file-based and UI configurations.

There is a discussion here, but no solution yet: 
https://community.openhab.org/t/update-to-openhab-3-1-0-breaks-homematic-binding/124095/3

**Comment by MHerbst** [comment]
I have probably solved the problem and created a test version of the binding: https://github.com/MHerbst/openhab-addons-test

Don't mind the version number. It will work with 3.1.

**Comment by s0170071** [comment]
confirmed, it works. Thank you.

Steps to do it: 
1. remove homematic binding from openhab (uninstall it with the gui)
2. drop the jar file from the link above into the addons folder /usr/share/openhab/addons. I have the RPI4 file system mounted under linux, but you can also use wget.
3. change ownership: navigate with the command prompt to the directory, then do (as root) `chown openhab:openhab org.openhab.binding.homematic-3.2.0-SNAPSHOT.jar`
4. restart openhab with `sudo systemctl restart openhab`
5. wait and watch the console. I did **not** have to flush any cache here. I did **not** have to manually re-install the binding from the console or gui.

**Comment by MHerbst** [comment]
Thanks for your feedback. I will enhance my installation description. I especially forgot point 3.

The re-install of the upnp feature is not necessary if any other binding uses upnp. But if the Homematic binding is the only one it will be automatically removed together with the HM binding.

**Comment by brainbug95** [comment]
Hi, I got the same problem but for homematic devices. HMIP is working. I use homegear as a gateway for homematic and CCU2 for HMIP devices. No docker.

`2021-07-06 21:24:58.942 [INFO ] [ommunicator.AbstractHomematicGateway] - HmGatewayInfo[id=HOMEGEAR,type=Homegear,firmware=0.8.0-3345,address=VBC4752592,rf=true,wired=false,hmip=false,cuxd=false,group=false]
2021-07-06 21:24:59.010 [INFO ] [ternal.communicator.client.RpcClient] - Interface 'CUXD' on gateway 'ccu2' not available, disabling support
2021-07-06 21:24:59.082 [INFO ] [ommunicator.AbstractHomematicGateway] - HmGatewayInfo[id=CCU,type=CCU2,firmware=2.57.5,address=OEQ1835612,rf=true,wired=false,hmip=true,cuxd=false,group=true]
2021-07-06 21:25:00.432 [ERROR] [very.HomematicDeviceDiscoveryService] - null
java.lang.NullPointerException: null
        at org.openhab.binding.homematic.internal.type.HomematicTypeGeneratorImpl.generate(HomematicTypeGeneratorImpl.java:173) ~[?:?]
        at org.openhab.binding.homematic.internal.handler.HomematicBridgeHandler.onDeviceLoaded(HomematicBridgeHandler.java:330) ~[?:?]
        at org.openhab.binding.homematic.internal.communicator.AbstractHomematicGateway.loadAllDeviceMetadata(AbstractHomematicGateway.java:415) ~[?:?]
        at org.openhab.binding.homematic.internal.discovery.HomematicDeviceDiscoveryService.lambda$0(HomematicDeviceDiscoveryService.java:209) ~[?:?]
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515) [?:?]
        at java.util.concurrent.FutureTask.run(FutureTask.java:264) [?:?]
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304) [?:?]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) [?:?]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) [?:?]
        at java.lang.Thread.run(Thread.java:829) [?:?]
`
I tried using the provided binding version 3.2.0.202107011418 but with the same result.

I completely downgraded the openhab installation to 3.02 again which makes eveything work again.

**Comment by s0170071** [comment]
oh, I forgot to mention that I did try to downgrade, failed and then 
apt-get removed openhab and reinstalled it. maybe that flushed a cache 🤔

also make sure you uninstall the old binding in the gui. and reboot. 
sometimes it takes more than one reboot.

Am 6. Juli 2021 21:33:50 schrieb Christian Weber ***@***.***>:

**Comment by MHerbst** [comment]
@brainbug95 This is definitely a different problem. The NPE occurs at another position in the code, and maybe it is related to wired devices. This can be related to a change in OH Core. Please create a new issue for this problem. Is it correct that you have defined two bridges? One for Homegear and one for the CCU? I will probably need more detailed log output to found the reason for this problem.

**Comment by brainbug95** [comment]
@s0170071 thanks for the hint. I made it a habit of rebooting the machine after any openhab updates/upgrades. I have had some weird caching effects in the past and rebooting was faster than manually clearing caches. 

@MHerbst I will create an new issue with more debug logging enabled once I find some time to do it. 
Yes I use two bridges for historic reasons. When I started with openhab there was no HMIP. I bought a Homematic Lan Gateway and connected it to homegear which is connected to openhab. Then there came HMIP and since the LAN gateway did not support HMIP Devices I got myself a CCU2. All HM-Devices are bound to the Homegear via LAN Gateway and HMIP Devices are bound to the CCU2. Worked like charm all versions >= 3.0.2.
